src/components/MetadataPanel/index.tsx

```diff
@@ -0,0 +1,151 @@
+
+import styles from "./style.css";
+
+interface MetadataPanelProps {
+    selectedAgent: AgentMetadata;
+    uiDisplayData: AgentDisplayNode[];
+}
+
+const MetadataPanel: React.FC<MetadataPanelProps> = ({
+    selectedAgent,
+    uiDisplayData,
+}) => {
+    const [panelExpanded, setPanelExpanded] = React.useState(false);
+
+    const agentSelected = !!selectedAgent.uniqueId;
+    const selectedAgentNotRendered =
+        selectedAgent.agentType >= uiDisplayData.length;
+
+    useEffect(() => {
+        if (!agentSelected) {
+            setPanelExpanded(false);
+        } else {
+            setPanelExpanded(true);
+        }
+    }, [selectedAgent.uniqueId]);
+
+    const displayValues: Record<keyof AgentMetadata, MetadataDisplay> = {
+        uniqueId: {
+            label: <span>Unique ID</span>,
+            value: () => formatFloatForDisplay(selectedAgent.uniqueId),
+        },
+        agentType: {
+            label: <span>Agent Type</span>,
+            value: () => uiDisplayData[selectedAgent.agentType].title,
+        },
+        position: {
+            label: <span>Position</span>,
+            value: () => {
+                const pos = selectedAgent.position;
+                return (
+                    <>
+                        <div>x = {formatFloatForDisplay(pos.x)}</div>
+                        <div>y = {formatFloatForDisplay(pos.y)}</div>
+                        <div>z = {formatFloatForDisplay(pos.z)}</div>
+                    </>
+                );
+            },
+        },
+        rotation: {
+            label: (
+                <>
+                    Rotation
+                    <br />
+                    (degrees)
+                </>
+            ),
+            value: () => {
+                const rot = selectedAgent.rotation;
+                return (
+                    <>
+                        <div>x = {formatFloatForDisplay(rot.x)}</div>
+                        <div>y = {formatFloatForDisplay(rot.y)}</div>
+                        <div>z = {formatFloatForDisplay(rot.z)}</div>
+                    </>
+                );
+            },
+        },
+        radius: {
+            label: <span>Radius</span>,
+            value: () => formatFloatForDisplay(selectedAgent.radius),
+        },
+    };
+
+    const metadataRows = useMemo(() => {
+        const keys = Object.keys(selectedAgent) as (keyof AgentMetadata)[];
+        return keys.map((key, index) => (
+            <React.Fragment key={key}>
+                <div className={styles.row}>
+                    <div className={styles.key}>{displayValues[key].label}</div>
+                    <div className={styles.value}>
+                        {displayValues[key].value()}
+                    </div>
+                </div>
+                {index < keys.length - 1 && (
+                    <Divider className={styles.divider} />
+                )}
+            </React.Fragment>
+        ));
+    }, [selectedAgent, selectedAgentNotRendered]);
```

Comment:
why is `selectedAgentNotRendered` in the requirements list? it doesn't seem like it's being used here. And I don't think you need useMemo here because this is going to get recalculated anyway every time the props change (ie, you're not saving any re-renders)

Reply:
Ah you're right!